### PR TITLE
expose privatelink status field

### PR DESCRIPTION
**Pull request by petrosagg**
### Motivation

This PR exposes the status field of the k8s PrivateLink resource (introduced in https://github.com/MaterializeInc/cloud/pull/6383)  to the materialize database. This API will be used to implement `VALIDATE CONNECTION` for privatelink connections and also to expose privatelink status as a table. None of these two are included in this PR.

In the process I refactored both Secret and CloudResource traits to follow the same pattern of a `FooController` trait that can mutate the managed objects and a `FooReader` trait that only has a read-only view on the resources. I made `FooController` to be a supertrait of `FooReader`.

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  -

### CODE REVIEW

File: src/secrets/src/lib.rs [hunk 1]
```diff
@@ -85,18 +85,14 @@ use mz_repr::GlobalId;
...
     /// contents.
     async fn ensure(&self, id: GlobalId, contents: &[u8]) -> Result<(), anyhow::Error>;
 
     /// Deletes the specified secret.
     async fn delete(&self, id: GlobalId) -> Result<(), anyhow::Error>;
 
-    /// Lists known secrets. Unrecognized secret objects do not produce an error
-    /// and are ignored.
-    async fn list(&self) -> Result<Vec<GlobalId>, anyhow::Error>;
-
```
Comment (benesch) [inline src/secrets/src/lib.rs:99]: Having `list` be part of the `SecretsController` API was intentional. While it's technically a read-only operation, the idea was that the controller contains operations that are only relevant if you are managing secrets, and the `SecretsReader` contains the single operation (reading a secret) that is relevant if all you need is to determine the contents of secrets whose IDs you are told. 

I'm not seeing a good reason to change this! Keeping the `SecretsReader` as simple as possible makes it easier to do something like a `CachingSecretsReader` as discussed here: https://materializeinc.slack.com/archives/CL68GT3AT/p1689805092813149?thread_ts=1689797856.040739&cid=CL68GT3AT

File: src/cloud-resources/src/lib.rs [hunk 2]
```diff
@@ -143,8 +146,19 @@ pub trait CloudResourceController: Debug + Send + Sync {
...
+pub trait CloudResourceReader: Debug + Send + Sync {
+    /// Reads the specified `VpcEndpoint` Kubernetes object.
+    async fn read(&self, id: GlobalId) -> Result<VpcEndpointStatus, anyhow::Error>;
+
     /// Lists existing `VpcEndpoint` Kubernetes objects.
-    async fn list_vpc_endpoints(&self) -> Result<BTreeSet<GlobalId>, anyhow::Error>;
+    async fn list_vpc_endpoints(
+        &self,
+    ) -> Result<BTreeMap<GlobalId, VpcEndpointStatus>, anyhow::Error>;
 }
```
[the remark this answers is not in the document]
Comment (petrosagg) [reply src/cloud-resources/src/lib.rs:162]: > What does having two traits here facilitate?

Nothing concrete other than in my eyes they are doing the same thing and therefore should look identical.

> we don't need to ship off support for reading VPC endpoints to lots of places in the codebase.

It's true that not many places will use the VPC endpoint controller but whatever object ends up being responsible for reading them will end up inside the `ConnectionContext` struct which is threaded in a bunch of places.

That said I'm a guest in this part of the codebase so I'm happy to follow whatever pattern is desired. Are you ok with putting the full `KubernetesOchestrator` in `ConnectionContext`?

File: src/secrets/src/lib.rs (hunk 1 above)
Comment (petrosagg) [reply src/secrets/src/lib.rs:99]: Happy to leave it as is, see my other comment

File: src/cloud-resources/src/lib.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (benesch) [reply src/cloud-resources/src/lib.rs:162]: Ohh, sorry, I see! I didn't realize that connection validation _only_ had access to the connection context. I had naively assumed it had access to the whole adapter state, including the `KubernetesOrchestrator`.

**Review by benesch — COMMENTED** [review]
Ok, thanks for the explainer. This makes sense to me now! I am 👍🏽 on `SecretsController` being a supertrait of `SecretsReader` and similarly for the `CloudResourceController`.

However, I still think it'd be worthwhile to keep the `list` operations in the `*Controller` traits instead of the `*Reader` traits. The semantics of a `CachingReader` are much clearer, IMO, if the `list` operation is not part of that trait. Does that sound good to you, too? Is there a better name than "Reader" for the traits? The distinction I wanted to draw was between "management" and "users" rather than "write" and "read".

**Comment by benesch** [comment]
Here's the other in-flight PR that this PR would conflict with as written: https://github.com/MaterializeInc/materialize/pull/20691.

**Comment by petrosagg** [comment]
Sounds good! I'll move the `list` to the `*Controller` traits

>  Is there a better name than "Reader" for the traits?

`SecretsLookup` is the only thing I came up with but `Reader` is ok too. Any preference?

**Comment by benesch** [comment]
> > Is there a better name than "Reader" for the traits?
> 
> `SecretsLookup` is the only thing I came up with but `Reader` is ok too. Any preference?

I guess I have the weakest of preferences for `SecretReader` but `SecretLookup` sounds fine to me too.

**Review by benesch — APPROVED** [review]

File: src/orchestrator-kubernetes/src/cloud_resource_controller.rs [hunk 3]
```diff
@@ -81,20 +83,62 @@ impl CloudResourceController for KubernetesOrchestrator {
...
+        let objects = self.vpc_endpoint_api.list(&ListParams::default()).await?;
+        let mut endpoints = BTreeMap::new();
+        for object in objects {
+            let maybe_id = object
+                .name_any()
+                .split_once('-')
+                .and_then(|(_, id_str)| GlobalId::from_str(id_str).ok());
+            let id = match maybe_id {
+                Some(id) => id,
+                // Ignore any whom's name can't be parsed into a GlobalId
```
Comment (sploiselle) [inline src/orchestrator-kubernetes/src/cloud_resource_controller.rs:98]: nit: `// Ignore any object whose name can't be parsed as a GlobalId`

File: src/orchestrator-kubernetes/src/cloud_resource_controller.rs [hunk 4]
```diff
@@ -81,20 +83,62 @@ impl CloudResourceController for KubernetesOrchestrator {
...
+}
+
+#[async_trait]
+impl CloudResourceReader for KubernetesOrchestrator {
+    async fn read(&self, id: GlobalId) -> Result<VpcEndpointStatus, anyhow::Error> {
+        self.resource_reader.read(id).await
+    }
+}
+
+/// Reads secrets managed by a [`KubernetesOrchestrator`].
```
Comment (sploiselle) [inline src/orchestrator-kubernetes/src/cloud_resource_controller.rs:119]: This name sounds more general than only reading secrets--typo or something I'm overlooking?

File: src/cloud-resources/src/crd/vpc_endpoint.rs [hunk 5]
```diff
@@ -44,9 +44,38 @@ pub mod v1 {
...
 
-    #[derive(Clone, Debug, Deserialize, Serialize, JsonSchema)]
+    #[derive(Clone, Debug, Default, Deserialize, Serialize, JsonSchema, PartialEq)]
     #[serde(rename_all = "camelCase")]
-    pub struct VpcEndpointStatus {}
+    pub struct VpcEndpointStatus {
+        // This will be None if the customer hasn't allowed our principal, got the name of their
+        // VPC Endpoint Service wrong, or we've otherwise failed to create the VPC Endpoint.
+        pub vpc_endpoint_id: Option<String>,
+        pub state: Option<VpcEndpointState>,
```
Comment (sploiselle) [inline src/cloud-resources/src/crd/vpc_endpoint.rs:53]: Are there any dependencies between the endpoint ID and the state? e.g. can you have the state without the ID? Having some comment here or an assertion on these relationships would make thing more grokkable.

File: src/orchestrator-kubernetes/src/cloud_resource_controller.rs [hunk 6]
```diff
@@ -81,20 +83,62 @@ impl CloudResourceController for KubernetesOrchestrator {
...
+#[derive(Debug)]
+pub struct KubernetesResourceReader {
+    vpc_endpoint_api: Api<VpcEndpoint>,
+}
+
+impl KubernetesResourceReader {
+    /// Constructs a new Kubernetes secrets reader.
```
Comment (sploiselle) [inline src/orchestrator-kubernetes/src/cloud_resource_controller.rs:126]: again, unsure about the relationship between `Resource` and `secrets`

**Review by sploiselle — APPROVED** [review]
Code LGTM, but have a few questions about the doc strings, and the structure of `VpcEndpointStatus` feels a little opaque to me (though I am unfamiliar with the idioms in this part of the code base, so very possible no one else would be confused).

File: src/orchestrator-kubernetes/src/cloud_resource_controller.rs (hunk 4 above)
Comment (petrosagg) [reply src/orchestrator-kubernetes/src/cloud_resource_controller.rs:119]: It was a typo from copy pasta :see_no_evil:

File: src/cloud-resources/src/crd/vpc_endpoint.rs (hunk 5 above)
Comment (petrosagg) [reply src/cloud-resources/src/crd/vpc_endpoint.rs:53]: These structs are merely copies of the structs maintained by the cloud team here https://github.com/MaterializeInc/cloud/blob/main/mzcloud-environment-controller/src/controller/vpc_endpoint.rs#L116-L139

My understanding is that this struct expresses the possibility that k8s can for whatever reason not have a state associated with an endpoint even though it exists.

File: src/cloud-resources/src/crd/vpc_endpoint.rs [hunk 7]
```diff
@@ -44,9 +44,38 @@ pub mod v1 {
...
+        Deleted,
+        Deleting,
+        Expired,
+        Failed,
+        // Customer has approved the connection. It should eventually move to Available.
+        Pending,
+        // Waiting on the customer to approve the connection.
+        PendingAcceptance,
+        Rejected,
+        Unknown,
```
[the remark this answers is not in the document]
Comment (petrosagg) [reply src/cloud-resources/src/crd/vpc_endpoint.rs:77]: `None` represents the state not being present in k8s at all and `Unknown` represents that we explicitly set the state to that so the reason it's unknown is not due to k8s. For users this doesn't matter and for both cases I just say "unknown" as the result of `VALIDATE CONNECTION`.

Justin from the cloud team is working on making these states a bit more useful by including additional context. Once that work lands we can change the mz side to reflect that and expose more information

File: src/cloud-resources/src/crd/vpc_endpoint.rs (hunk 5 above)
Comment (sploiselle) [reply src/cloud-resources/src/crd/vpc_endpoint.rs:53]: Ah, this was the context I was missing. Ty.

**Pull request merged**